sematic/resolvers/worker.py

```diff
@@ -70,14 +70,22 @@ def _get_input_kwargs(
     return kwargs
 
 
-def _fail_run(run: Run):
+def _fail_run(run: Run, e: BaseException) -> None:
     """
     Mark run as failed.
     """
     run.future_state = FutureState.FAILED
     run.failed_at = datetime.datetime.utcnow()
-    if run.exception is None:
-        run.exception = format_exception_for_run()
+
+    if run.exception_metadata is not None:
+        # if the run already has an exception marked on it, then it's the innermost cause
+        # of the failure; any other exception generated afterwards is done so while trying
+        # to handle the failure
+        logger.warning("Got exception while handling run failure", exc_info=e)
+    else:
+        # this means the exception probably happened in the Resolver code
+        run.exception_metadata = format_exception_for_run(e)
```

Comment:
Done.

I use 2 principles:

- If one of the branches is relatively very short compared to the other, it should go first. This way the reader can quickly read it and continue reading the alternative. If the first branch is relatively very large, there is the risk of the reader forgetting the if condition by the time they read the first branch, and having to go back to the if to read it, disrupting the flow.

- If the if condition obviously highlights / hints that one of the branches is critical or controversial in any way, then the user will be anxious to read that branch, and it should come first. Otherwise, they will not focus when reading the "less-important" first branch and miss things.

None of these 2 situations especially applies here, so I went with your preference.